### DISCUSSION TITLE
Workspaces in the npm CLI

### DISCUSSION THREAD

**Discussion by MylesBorins** [npm/feedback#5] [CLI] [answered]
Lots of folks work with monorepos and it would be great to have more official tooling to support workflows

**Comment by ethomson** [comment]
This is something that we're looking at adding support for in npm v7.

/cc @darcyclarke @ruyadorno

**Comment by MylesBorins** [comment]
I think a big question here is if we want to have one larger tracking issue or if we want to break down a number of specific workspace features we would like to see landed.

Likely it will be a bit of both. A meta tracking issue for considering a baseline of workspaces complete as well as some more scoped issues for specific workspace features.

**Comment by targos** [comment]
Is this discussion about adding features that exist for example in [Lerna](https://lerna.js.org/)?

**Reply by MylesBorins** [reply]
Basics have already been covered in https://github.com/npm/rfcs/pull/103 and will ship with 7.0.0. I was imagining this discussion could focus on broader support we should consider including. This might branch into multiple other discussions

**Comment by wesleytodd** [comment, marked as answer]
This has been implemented in npm 7, I have even used the beta on a project with it.  See the RFC: https://github.com/npm/rfcs/pull/103

**Reply by MylesBorins** [reply]
@wesleytodd my understanding is that the basics of workspaces has been implemented but there is still room for more features. I was imagining this discussion could be a way to kick off "what is missing" in the current implementation

**Reply by wesleytodd** [reply]
Cool, I will re-share my feedback on this then.  Thanks for the clarification!

**Reply by MylesBorins** [reply]
Thanks for sharing the link to the initial RFC. I didn't know where to look to find what was already implemented.

@ruyadorno pointed me towards https://github.com/npm/rfcs/pull/117 for some of the next steps being discussed

**Comment by wesleytodd** [comment]
One big gap with workspaces in all existing tools is they focus strongly on monorepos.  But this is a rather solved space and IMO the much more common and more impactful feature addition would be workspaces which are multi-repo.  This is almost all of the important and long standing projects on the public registry, and also not covered under any existing tooling today and is more difficult to manage than monorepos.

**Reply by MylesBorins** [reply]
So the ask here I'm guessing is for something that could replace or improve `npm link`?

**Reply by wesleytodd** [reply]
Maybe.  I think it is more complicated than that, but if we started with something which let you define a set to all just `npm link` it might even be fine.  The point is, workspaces as they are today require a specific file structure (ws root as the parent dir).  This means for my typical setup with express modules for example:

```
- projects
 | - express
 | - pillarjs
 |  | - router
 |  | - path-to-regexp
 | - other-unrelated-projects
```

I would need to put a `package.json` with a `workspaces` key in `projects`, but then if I switched context to `other-unrelated-projects`, I would blow away the one to replace it with the other.  This is pretty ineffective.

I would much rather have something like VSCode workspaces where it is just a definition file somewhere and the paths do not require nested structure.  I have some code prototyping this out with an import map loader, but the team decided to go the yarn feature parity route and the import maps discussion was frustrating, so I moved on to other things for the time being.

**Reply by MylesBorins** [reply]
I'd definitely be interested in hearing more about how this could be accomplished with import maps, as you know I very much like that technology.

Would you imagine the role of npm in that case would be generating the import maps which would then be provided to any supporting runtime?

**Reply by wesleytodd** [reply]
Yep exactly.  The demo I had used arborist and just walked the tree to generate the map.  I am not sure this will make sense out of context (this was inside an experiment I was doing to do managed configurations for projects, so the `Plugin` stuff is wrapper code), but here is the code I was using at the time:

```javascript
'use strict'
const Arborist = require('@npmcli/arborist')
const globby = require('globby')
const path = require('path')
const fs = require('fs-extra')
const { flatten } = require('array-flatten')

module.exports.init = (Plugin) => {
  return class ImportMap extends Plugin {
    constructor () {
      super()
      this.name = 'space/importmap'
      this.importMap = {
        imports: {},
        scopes: {}
      }
    }

    static LOADER_TMPL = `import module from 'module'

const importmap = %%IMPORT_MAP%%

const orig = module._resolveFilename
module._resolveFilename = (request, parent, isMain, options) => {
  // Check scopes
  let rent = parent
  while (rent) {
    if (importmap.scopes[rent.path] && importmap.scopes[rent.path][request]) {
      return importmap.scopes[rent.path][request]
    }
    rent = rent.parent
  }

  // Check imports
  if (importmap.imports[request]) {
    return importmap.imports[request]
  }

  // Fall back to filesystem lookup
  return orig.call(module, request, parent, isMain, options)
}
`
    async configure (overrides) {
      return {
        loaderTmpl: overrides.loaderTmpl || ImportMap.LOADER_TMPL,
        loaderPath: overrides.loaderPath || this.TMPL`${'rootProject.root'}/loader.mjs`
      }
    }

    async load () {
      // Reify top tree
      await this.installTree(this.rootProject.root)

      // Write import map
      const out = this.get('loaderTmpl').replace('%%IMPORT_MAP%%', JSON.stringify(this.importMap, '', '  '))
      await fs.writeFile(this.get('loaderPath'), out)
    }

    async installTree (dir) {
      const arb = new Arborist({
        path: dir,
        fullMetadata: true
      })
      const tree = await arb.buildIdealTree()

      // Add workspace links
      if (Array.isArray(tree.package.workspaces)) {
        const pPaths = tree.package.workspaces.map((glob) => {
          return globby(path.resolve(arb.path, glob), {
            expandDirectories: false,
            onlyDirectories: true
          })
        })
        const paths = await Promise.all(pPaths)

        // Create flattened subtrees
        await Promise.all(flatten(paths).map(async (p) => {
          const t = await this.installTree(path.resolve(arb.path, p))

          // eslint-disable-next-line
          new Arborist.Link({
            name: t.name,
            parent: tree,
            target: t
          })
        }))
      }

      // Reify tree
      await arb.reify()

      // Add children to the import map
      await this.addToMap(tree)

      return tree
    }

    async addToMap (node) {
      // Build the import map
      if (node.isRoot) {
        this.importMap.imports[node.name] = path.join(node.path, node.package.main || 'index.js')
        this.importMap.imports[`${node.name}/`] = node.path
      } else {
        if (!node.package._id) {
          // Hack back in package.json?
          console.log('didnt load full metadata', node.package)
          return
        }
        this.importMap.scopes[node.parent.path] = this.importMap.scopes[node.parent.path] || {}
        this.importMap.scopes[node.parent.path][node.name] = path.join(node.path, node.package.main || 'index.js')
        this.importMap.scopes[node.parent.path][`${node.name}/`] = node.path
      }

      // Recurse tree
      for (const [, edge] of node.edgesOut) {
        if (edge.to && edge.to.isLink) {
          // debugger
          // @TODO what to do here?
          // await this.installTree(edge.to.target.path)
        } else if (edge.to) {
          await this.addToMap(edge.to)
        } else {
          // @TODO hit for optional and unfulfilled deps
          // console.log(edge.type)
        }
      }
    }

    async onExec ([cmd, args, opts]) {
      if (!cmd.endsWith('bin/node')) {
        return [cmd, args, opts]
      }

      // Set loader flag
      args.unshift(
        '--experimental-modules',
        '--loader', path.join(this.root, 'loader.mjs')
      )

      return [cmd, args, opts]
    }
  }
}
```

EDIT: Reading this again I remember how hacky it is :P
